openproblems/tasks/spatial_decomposition/_utils.py

```diff
@@ -1,16 +1,61 @@
-from anndata import AnnData
+from typing import Tuple
 
+import anndata as ad
 import numpy as np
+import pandas as pd
 
 
-def obs_means(adata: AnnData, cluster_key: str) -> AnnData:
+def merge_sc_and_sp(
+    adata_sc: ad.AnnData,
+    adata_sp: ad.AnnData,
+    batch_key: str = "modality",
+) -> ad.AnnData:
+
+    for k in adata_sp.obsm.keys():
+        if isinstance(adata_sp.obsm[k], pd.DataFrame):
+            n_col = adata_sp.obsm[k].shape[1]
+            n_row = adata_sc.shape[0]
+            df = pd.DataFrame(
+                np.zeros((n_row, n_col)),
+                columns=adata_sp.obsm[k].columns.copy(),
+                index=adata_sc.obs_names.copy(),
+            )
+            adata_sc.obsm[k] = df
+
+    # merge single cell and spatial data
+    adata_merged = ad.concat(
```

Comment:
Maybe try using `ad.concatenate` since `ad.concat` is still experimental in 0.7.x